pkgs/dart_tooling_mcp_server/lib/src/mixins/dtd.dart

```diff
@@ -283,6 +284,65 @@ base mixin DartToolingDaemonSupport on ToolsSupport {
     );
   }
 
+  /// Retrieves the Flutter widget tree from the currently running app.
+  ///
+  /// If more than one debug session is active, then it just uses the first one.
+  ///
+  // TODO: support passing a debug session id when there is more than one debug
+  // session.
+  Future<CallToolResult> widgetTree(CallToolRequest request) async {
+    return _callOnVmService(
+      callback: (vmService) async {
+        final vm = await vmService.getVM();
+        final isolateId = vm.isolates!.first.id;
+        final groupId = 'dart-tooling-mcp-server';
+        const inspectorExtensionPrefix = 'ext.flutter.inspector';
+        try {
+          final result = await vmService.callServiceExtension(
+            '$inspectorExtensionPrefix.getRootWidgetTree',
+            isolateId: isolateId,
+            args: {
+              'groupName': groupId,
+              // TODO: consider making these configurable or using defaults that
+              // are better for the LLM.
+              'isSummaryTree': 'true',
+              'withPreviews': 'true',
+              'fullDetails': 'true',
+            },
+          );
+          final tree = result.json?['result'];
+          if (tree == null) {
+            return CallToolResult(
+              content: [
+                TextContent(
+                  text:
+                      'Could not get Widget tree. '
+                      'Unexpected result: ${result.json}.',
+                ),
+              ],
+            );
+          }
+          return CallToolResult(content: [TextContent(text: tree.toString())]);
+        } catch (e) {
+          return CallToolResult(
+            isError: true,
+            content: [
+              TextContent(
+                text: 'Unknown error or bad response getting widget tree:\n$e',
+              ),
+            ],
+          );
+        } finally {
+          await vmService.callServiceExtension(
+            '$inspectorExtensionPrefix.disposeGroup',
+            isolateId: isolateId,
+            args: {'objectGroup': groupId},
+          );
```

Comment:
@elliette is cleaning up this object group necessary? Or do you think we should re-use this object group for all uses of this tool and clean up when the server gets shut down? I'm not super familiar with why we use these object groups in the Inspector.

Reply:
Hmm I'm also not familiar with the object group lifecycle. I see this comment: https://github.com/flutter/flutter/blob/296df33be866d0cd8387524e288d9b280e725795/packages/flutter/lib/src/widgets/widget_inspector.dart#L924-L927

But I'm not sure if how we should be creating / disposing object groups is documented anywhere. I need to do more investigation